librz/util/buf.c

```diff
@@ -1283,15 +1283,68 @@ RZ_API void rz_buf_set_overflow_byte(RZ_NONNULL RzBuffer *b, ut8 Oxff) {
  * \param size ...
  *
  * WARNING: this function should be used with care because it may allocate the
- * entire buffer in memory. Consider using the rz_buf_read* APIs instead and
+ * entire buffer in memory. Consider using the rz_buf_read* APIs or rz_buf_fwd_scan API instead and
  * read only the chunks you need.
  */
-RZ_DEPRECATE RZ_API RZ_BORROW const ut8 *rz_buf_data(RZ_NONNULL RzBuffer *b, ut64 *size) {
+RZ_DEPRECATE RZ_API RZ_BORROW ut8 *rz_buf_data(RZ_NONNULL RzBuffer *b, ut64 *size) {
 	rz_return_val_if_fail(b, NULL);
 
 	return get_whole_buf(b, size);
 }
 
+/**
+ * \brief Scans buffer linearly in chunks calling \p fwd_scan for each chunk.
+ *
+ * \param b RzBuffer to read
+ * \param start Start address
+ * \param amount Amount of bytes to read
+ * \param fwd_scan Function to call for each chunk
+ * \param user User data to pass to fwd_scan
+ * \return Number of bytes read
+ */
+RZ_API ut64 rz_buf_fwd_scan(RZ_NONNULL RzBuffer *b, ut64 start, ut64 amount, RZ_NONNULL RzBufferFwdScan fwd_scan, RZ_NULLABLE void *user) {
+	rz_return_val_if_fail(b && fwd_scan, 0);
+	if (!amount) {
+		return 0;
+	}
+	if (b->methods->get_whole_buf) {
+		ut64 sz;
+		const ut8 *buf = b->methods->get_whole_buf(b, &sz);
+		if (buf) {
+			if (sz <= start) {
+				return 0;
+			}
+			return fwd_scan(buf + start, RZ_MIN(sz - start, amount), user);
+		}
```

Comment:
is this correct? because if buf is NULL i expect an error.

Reply:
this is necessary because the if the buffer is an wrapper to an `RzIO` file descriptor, `get_whole_buffer` depends if the underlying io plugin implements `get_buf`.